### PR TITLE
WIP Release/v0.13.11

### PR DESCRIPTION
**Pull request by michaloo**
tl;dr

PR adds custom Error classes which when thrown or passed to promise rejection will be treated differently - it won't be sent to sentry but a datadog metric `connector.transient_error` with two tags `error_name` and `error_message` will be instrumented to see the rate of specific error and optionally bubble up to an internal alert or customer facing information.

full changelog:
* this release brings bigger changes to error handling:
  - it cleans up a little middleware stack including smart-notifier errors
  - it introduces two types of errors - `unhandled error` which is handled the same as till now, and `transient error` which won't be pushed to sentry, but only instrumented in datadog
  - it deprecates dedicated smartNotifierErrorMiddleware
  - smartNotifierHandler in case of error behaves like notifHandler and pass the error down the middleware stack
* added `timeout` option to `Hull.Connector` constructor to control the timeout value
* upgrades `raven` library
* flow types fixes

### CODE REVIEW

**Comment by unity** [comment]
@michaloo can you document usage examples in `readme.md`

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

File: README.md [hunk 1]
```diff
@@ -516,244 +372,65 @@ app.post('/request', (req, res) => {
...
-```javascript
-function userUpdateHandler(ctx, messages = []) {
-  ctx.smartNotifierResponse.setFlowControl({
-    type: 'next',
-    in: 1000
-  });
-  return Promise.resolve();
-}
-```
+Hull Node promotes using [SuperAgent](http://visionmedia.github.io/superagent/) as a core HTTP client. We provide two plugins to add more instrumentation over the requests.
```
Comment (unity) [inline README.md]: @michaloo missing the actual plugins descriptions here

**Comment by michaloo** [comment]
@unity sure, still doing major moves between `README.md` (focused around how-to and getting started and examples) and `API.md` (precise and exact information about tools you already know how to use). Thanks for fixes.

File: API.md [hunk 2]
```diff
@@ -0,0 +1,1246 @@
...
+**Parameters**
+
+-   `HullClient` **HullClient** Hull Client - the version exposed by this library comes with HullClient argument bound
+-   `options` **[Object][57]** 
+    -   `options.hostSecret` **[string][58]** The ship hosted secret - consider this as a private key which is used to encrypt and decrypt `req.hull.token`. The token is useful for exposing it outside the Connector &lt;-> Hull Platform communication. For example the OAuth flow or webhooks. Thanks to the encryption no 3rd party will get access to Hull Platform credentials.
+    -   `options.clientConfig` **[Object][57]** Additional config which will be passed to the new instance of Hull Client (optional, default `{}`)
+
+Returns **[Function][61]** 
+
+## Types
```
Comment (unity) [inline API.md]: Weird to have Types in the middle of the rest.

File: API.md [hunk 3]
```diff
@@ -0,0 +1,1246 @@
...
+where first element is an old value and second element is the new value.
+This object contain information about changes on one or multiple attributes (that's thy attributes and changes are plural).
+
+Type: {}
+
+### THullConnector
+
+Connector (also called ship) object with settings, private settings and manifest.json
+
+Type: {id: [string][58], updated_at: [string][58], created_at: [string][58], name: [string][58], description: [string][58], tags: [Array][64]&lt;[string][58]>, manifest: [Object][57], settings: [Object][57], private_settings: [Object][57], status: [Object][57]}
```
Comment (unity) [inline API.md]: Formatting broken

File: API.md [hunk 4]
```diff
@@ -0,0 +1,1246 @@
...
+General type for THullUser and THullAccount
+
+Type: (THullUser | THullAccount)
+
+### THullReqContext
+
+Context added to the express app request by hull-node connector sdk.
+Accessible via `req.hull` param.
+
+Type: {requestId: [string][58], config: [Object][57], token: [string][58], client: [Object][57], ship: THullConnector, connector: THullConnector, hostname: [string][58], options: [Object][57], connectorConfig: [Object][57], segments: [Array][64]&lt;THullSegment>, cache: [Object][57], metric: [Object][57], enqueue: [Function][61], helpers: [Object][57], service: [Object][57], shipApp: [Object][57], message: [Object][57]?, notification: [Object][57], smartNotifierResponse: [Object][57]?}
```
Comment (unity) [inline API.md]: Formatting broken

File: API.md [hunk 5]
```diff
@@ -0,0 +1,1246 @@
...
+
+**Examples**
+
+```javascript
+req.hull.helpers.updateSettings({ newSettings });
+```
+
+Returns **[Promise][62]** 
+
+## Context
```
Comment (unity) [inline API.md]: missing introductory paragraph with a code snippet of all the possible items? like on https://github.com/hull/hull-node/blob/develop/API.md#hullconnector

File: API.md [hunk 6]
```diff
@@ -0,0 +1,1246 @@
...
+
+This is a supplement method which calls `app.listen` internally and also terminates instrumentation of the application calls.
+
+**Parameters**
+
+-   `app` **express** expressjs application
+
+Returns **http.Server** 
+
+## Errors
```
Comment (unity) [inline API.md]: Is it possible to reorder content?
I'd do:
- HullConnector
- Hull.Middleware
- Context
- Infra
- Utils
- Helpers (btw, seems redundant to have Utils & Helpers - but that's probably a refactor)
- Types
- Errors

File: API.md [hunk 7]
```diff
@@ -0,0 +1,1246 @@
...
+  })
+  .catch((error) => {
+    // error.constructor.name can be ConfigurationError, RateLimitError coming from `ok` handler above
+    // or TransientError coming from logic applied by `superagentErrorPlugin`
+  })
+```
+
+Returns **[Function][61]** function to use as superagent plugin
+
+### superagentInstrumentationPlugin
```
Comment (unity) [inline API.md]: Seems weird to have this and a `Infra.InstrumentationAgent` totally unrelated, with completely different usage. But that's probably a refactor.

File: API.md [hunk 8]
```diff
@@ -0,0 +1,1246 @@
...
+
+```javascript
+const { Instrumentation } = require("hull/lib/infra");
+
+const instrumentation = new Instrumentation();
+
+const connector = new Connector.App({ instrumentation });
+```
+
+### QueueAgent
```
Comment (unity) [inline API.md]: Do we want to document this or deprecate it and ignore it in the docs ?

File: API.md [hunk 9]
```diff
@@ -0,0 +1,1246 @@
...
+
+This is a method to request an extract of user base to be sent back to the Connector to a selected `path` which should be handled by `notifHandler`.
+
+**Parameters**
+
+-   `ctx` **[Object][57]** Hull request context
+-   `options` **[Object][57]**  (optional, default `{}`)
+    -   `options.segment` **[Object][57]**  (optional, default `null`)
+    -   `options.format` **[Object][57]**  (optional, default `json`)
+    -   `options.path` **[Object][57]**  (optional, default `batch`)
```
Comment (unity) [inline API.md]: i'd document `default: /batch` (with Slash) so people understand it's a URL

File: API.md [hunk 10]
```diff
@@ -0,0 +1,1246 @@
...
+req.hull.helpers.updateSettings({ newSettings });
+```
+
+Returns **[Promise][62]** 
+
+## Context
+
+### cache
+
+Cache available as `req.hull.cache` object
```
Comment (unity) [inline API.md]: Maybe a quick line on Usage and how it relates to `CacheAgent` would help here.

File: API.md [hunk 11]
```diff
@@ -0,0 +1,1246 @@
...
+
+**Parameters**
+
+-   `key` **[string][58]** 
+
+Returns **any** Promise
+
+### metric
+
+Metric agent available as `req.hull.metric` object
```
Comment (unity) [inline API.md]: Some simple code sample of usage would help

File: API.md [hunk 12]
```diff
@@ -0,0 +1,1246 @@
...
+### THullReqContext
+
+Context added to the express app request by hull-node connector sdk.
+Accessible via `req.hull` param.
+
+Type: {requestId: [string][58], config: [Object][57], token: [string][58], client: [Object][57], ship: THullConnector, connector: THullConnector, hostname: [string][58], options: [Object][57], connectorConfig: [Object][57], segments: [Array][64]&lt;THullSegment>, cache: [Object][57], metric: [Object][57], enqueue: [Function][61], helpers: [Object][57], service: [Object][57], shipApp: [Object][57], message: [Object][57]?, notification: [Object][57], smartNotifierResponse: [Object][57]?}
+
+**Properties**
+
+-   `requestId` **[string][58]**
```
Comment (unity) [inline API.md]: Is it possible to render it as raw JS ?

I feel here a format such as 

```js
type {
  requestId: string
}
```

would be much more explicit

File: API.md [hunk 13]
```diff
@@ -0,0 +1,1246 @@
...
+-   `id` **[string][58]** 
+-   `anonymous_id` **[Array][64]&lt;[string][58]>** 
+-   `email` **[string][58]** 
+-   `account` **{}** 
+
+## Utils
+
+General utilities
+
+### notifHandler
```
Comment (unity) [inline API.md]: Do we want to document this since it's deprecated ?

File: API.md [hunk 14]
```diff
@@ -0,0 +1,1246 @@
...
+  },
+  ship: ship //The entire Ship instance's config
+}
+```
+
+Returns **[Function][61]** OAuth handler to use with expressjs
+
+### smartNotifierHandler
+
+`smartNotifierHandler` is a next generation `notifHandler` cooperating with our internal notification tool. It handles Backpressure, throttling and retries for you and lets you adapt to any external rate limiting pattern.
```
Comment (unity) [inline API.md]: If we're deprecating `notifHandler` we don't want to refer to it here I guess. And we can rename `internal notification tool` to `Kraken`.

File: API.md (hunk 6 above)
Comment (michaloo) [reply API.md]: good one, I will set that order. Thanks for input.
